packages/file_selector/file_selector_platform_interface/lib/src/types/x_file/io.dart

```diff
@@ -0,0 +1,106 @@
+import 'dart:convert';
+import 'dart:io';
+import 'dart:typed_data';
+
+import './base.dart';
+
+/// A XFile backed by a dart:io File.
+class XFile extends XFileBase {
+  final File _file;
+  final String mimeType;
+  final DateTime _lastModified;
+  int _length;
+
+  final Uint8List _bytes;
+
+  /// Construct a XFile object backed by a dart:io File.
+  XFile(
+    String path, {
+    this.mimeType,
+    String name,
+    int length,
+    Uint8List bytes,
+    DateTime lastModified,
+  })  : _file = File(path),
+        _bytes = null,
+        _lastModified = lastModified,
+        super(path);
+
+  /// Construct an XFile from its data
+  XFile.fromData(
+    Uint8List bytes, {
+    this.mimeType,
+    String path,
+    String name,
+    int length,
+    DateTime lastModified,
+  })  : _bytes = bytes,
+        _file = File(path ?? ''),
+        _length = length,
+        _lastModified = lastModified,
+        super(path) {
+    if (length == null) {
+      _length = bytes.length;
+    }
+  }
+
+  @override
+  Future<DateTime> lastModified() {
+    if (_lastModified != null) {
+      return Future.value(_lastModified);
+    }
+    return _file.lastModified();
+  }
+
+  @override
+  void saveTo(String path) async {
+    File fileToSave = File(path);
+    await fileToSave.writeAsBytes(_bytes ?? (await readAsBytes()));
+    await fileToSave.create();
+  }
+
+  @override
+  String get path {
+    return _file.path;
+  }
+
+  @override
+  String get name {
+    return _file.path.split(Platform.pathSeparator).last;
+  }
+
+  @override
+  Future<int> length() {
+    if (_length != null) {
+      return Future.value(_length);
+    }
+    return _file.length();
+  }
+
+  @override
+  Future<String> readAsString({Encoding encoding = utf8}) {
+    if (_bytes != null) {
+      return Future.value(String.fromCharCodes(_bytes));
+    }
+    return _file.readAsString(encoding: encoding);
+  }
+
+  @override
+  Future<Uint8List> readAsBytes() {
+    if (_bytes != null) {
+      return Future.value(_bytes);
+    }
+    return _file.readAsBytes();
+  }
+
+  @override
+  Stream<Uint8List> openRead([int start, int end]) async* {
+    if (_bytes != null) {
+      final bytes = _bytes;
+      yield bytes.sublist(start ?? 0, end ?? bytes.length);
+    }
+    yield* _file
+        .openRead(start ?? 0, end)
+        .map((chunk) => Uint8List.fromList(chunk));
```

Comment:
Can you just 

```dart
return _file.openRead(start ?? 0, end);
```

?

Reply:
I think
```
yield* _file.openRead(start ?? 0, end);
```
is right. I'm uncertain on the .map(..) function, I think it is just converting [openRead(..)](https://api.dart.dev/stable/2.9.3/dart-io/File/openRead.html) List<int> to Uint8List.

`return` can't be used because I am now using async* to supply _bytes if they are available. Receive error:

`Can't return a value from a generator function (using the async* modifier)`

`yield*` from my understanding will return `openRead`'s stream

I got this here:
https://github.com/flutter/plugins/blob/00d85b068715254992b1c940f2c8697791bf1507/packages/image_picker/image_picker_platform_interface/lib/src/types/picked_file/io.dart#L32-L36

Reply:
Ah, this is because the function is `async *`. What about extracting the async* part of the code to its own function, and let the normal case of openRead (from path) be handled directly by dart:io?

Reply:
Sweet sounds good, I'll try this. This map(..) is still necessary?

Reply:
I can't remember why that was needed (x_x)